### PR TITLE
add tests for json-memfd print/parse

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
Also:

 - add a convenient utility macro for asserting `GError` message contents
 - fix a minor issue with the escaping in the printer code

### CODE REVIEW

File: src/common/test-jsonfds.c [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+test_memfd_json_error_cases (void)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(JsonObject) object = NULL;
+  gint fd;
+  gint r;
+
+  /* invalid json */
+  fd = memfd_create ("xyz", MFD_CLOEXEC | MFD_ALLOW_SEALING);
+  write (fd, "beh", 3);
```
Comment (Gundersanne) [inline src/common/test-jsonfds.c:504]: tests complain about this:
```
src/common/test-jsonfds.c:504:3: error: ignoring return value of ‘write’, declared with attribute warn_unused_result [-Werror=unused-result]
   write (fd, "beh", 3);
   ^~~~~~~~~~~~~~~~~~~~
src/common/test-jsonfds.c:514:3: error: ignoring return value of ‘write’, declared with attribute warn_unused_result [-Werror=unused-result]
   write (fd, "[]", 2);
```

File: src/common/test-jsonfds.c [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+  r = fcntl (fd, F_ADD_SEALS, F_SEAL_SHRINK | F_SEAL_GROW | F_SEAL_WRITE);
+  g_assert_cmpint (r, ==, 0);
+  object = cockpit_memfd_read_json (fd, &error);
+  cockpit_assert_error_matches (error, JSON_PARSER_ERROR, JSON_PARSER_ERROR_INVALID_BAREWORD, "*unexpected identifier*");
+  g_clear_error (&error);
+  close (fd);
+
+  /* valid json, but not an object */
+  fd = memfd_create ("xyz", MFD_CLOEXEC | MFD_ALLOW_SEALING);
+  write (fd, "[]", 2);
```
Comment (Gundersanne) [inline src/common/test-jsonfds.c:514]: ditto

File: src/common/cockpitjsonprint.c [hunk 3]
```diff
@@ -28,21 +28,27 @@
...
-  /* signed comparison: `c < ' '` will also catch all non-ascii characters */
-  return c < ' ' || c == '\\' || c == '"';
+  /* we escape:
+   *   - ascii controls
+   *   - backslash
+   *   - double quote
+   *   - ascii del
+   *   - all non-ascii
+   */
+  return c < ' ' || c == '\\' || c == '"' || c >= 0x7f;
```
Comment (Gundersanne) [inline src/common/cockpitjsonprint.c:40]: using `unsigned` makes it seem more explicit, so I like it

File: src/common/test-jsonfds.c [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+test_print_string_memory_safety (void)
+{
+  gchar characters[] = { '\n', ' ', 'a', '\\', '\"', 0xcc };
+  const gchar *escaped[] = { "\\u000a", " ", "a", "\\\\", "\\\"", "?" };
+  gchar buffer[MAX_GROUPS * MAX_REPS];
+  Helpers helpers;
+
+  helpers_init (&helpers);
+
+  for (gint n_groups = MIN_GROUPS; n_groups <= MAX_GROUPS; n_groups++)
```
Comment (Gundersanne) [inline src/common/test-jsonfds.c:266]: In general I'm not a fan of making tests/test content generated like this. It makes it more difficult to read.

What is the benefit of doing it in a procedural manner versus writing a big json string into your expected_buffer and calling `test_string`? The latter would make it much more explicit, like in your `test_print_numeric`.

**Review by Gundersanne — COMMENTED** [review]
I have some questions. I'm not sure if my review will be of that much worth here, but I'll give it a try :)

File: src/common/test-jsonfds.c (hunk 4 above)
Comment (allisonkarlitskaya) [reply src/common/test-jsonfds.c:267]: This is testing thousands (or millions?) of possible combinations...

File: src/common/test-jsonfds.c (hunk 1 above)
Comment (allisonkarlitskaya) [reply src/common/test-jsonfds.c]: argh.  I added an assert.

File: src/common/test-jsonfds.c (hunk 4 above)
Comment (Gundersanne) [reply src/common/test-jsonfds.c:267]: can you at least explain what it's doing in a bit more detail then?

Comment (allisonkarlitskaya) [reply src/common/test-jsonfds.c:253]: done.

File: src/common/cockpitmemfdread.c [hunk 5]
```diff
@@ -48,6 +54,9 @@ cockpit_memfd_read (int      fd,
       return NULL;
     }
 
+workaround:
+  ;
```
Comment (Gundersanne) [inline src/common/cockpitmemfdread.c:58]: why the weird line with just a `;` here?

**Review by Gundersanne — APPROVED** [review]
thanks :)

File: src/common/cockpitmemfdread.c (hunk 5 above)
Comment (allisonkarlitskaya) [reply src/common/cockpitmemfdread.c:58]: You can't have a variable declaration following a label, so this turns it into a statement.

**Pull request merged**